### PR TITLE
Fixing and improving stuff

### LINKED ISSUE #5746: Experimental weapons animation
One of the weapons has a turf animation that doesn't go away after the explosion. I assume its the stickybomb launcher

### PR DESCRIPTION
**Pull request by DeityLink**
- Fixed (maybe) a bug that caused small explosions animations to keep re-appearing on the same turfs when leaving the area and coming back.
- Explosions now affect furthest turfs first and closest ones last, which makes explosion obstacles as reliable as they were intended to be.
- Improved the EMP method. Allowing for station-wide EMPs that don't lag the server.
- Supermatter Shard delaminations now cause a station-wide EMP. The EMP ripples slowly from the epicenter to the outskirts.

Fixes #5746

### CODE REVIEW

**Comment by Probe1** [comment]
>Fixed (maybe) a bug that caused small explosions animations to keep re-appearing on the same turfs when leaving the area and coming back.

I figured that would never be fixed.

File: code/__HELPERS/unsorted.dm [hunk 1]
```diff
@@ -1407,3 +1407,107 @@ proc/find_holder_of_type(var/atom/reference,var/typepath) //Returns the first ob
...
+	var/movement_dir = NORTH
+	var/segment_length = 1
+
+	if(inward)
+		pointer_x = -max_range
+		pointer_y = -max_range
+		segment_length = max_range*2+1
+		segment = 1
+
+		for(var/sstep=max_steps-1;sstep>=0;sstep--)
```
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: ```for(var/sstep in max_steps-1 to 0)```
is the fastest form of this loop.

File: code/__HELPERS/unsorted.dm [hunk 2]
```diff
@@ -1407,3 +1407,107 @@ proc/find_holder_of_type(var/atom/reference,var/typepath) //Returns the first ob
...
+				if(EAST)
+					pointer_x++
+				if(SOUTH)
+					pointer_y--
+				if(WEST)
+					pointer_x--
+			if(draw_red)
+				sleep(1)
+	else
+		for(var/sstep=1;sstep<=max_steps;sstep++)
```
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: ```for(var/sstep in 1 to max_steps)```
is the fastest form for this loop

**Comment by DeityLink** [comment]
@RemieRichards I take note.

When I tested spiral_block() with a range of 200 and without sleep(1) it still turned the Z level red pretty much instantly. The main source of lag if I was to guess comes from looping through every atoms in every turfs of the list and applying ex_act/emp_act to them. Not much we can do to improve that part I'm afraid.

**Comment by RemieRichards** [comment]
actually the ```to``` keyword has some rather RIDICULOUS scaling power.
it performs barely better than standard ```i++``` iteration on low numbers, but performs soooooooooooooooooooo much better on higher numbers.
and I've never encountered cases where ```to``` is applicable and worse than ```i+```

tl;dr free optimisation with no downside.

**Comment by DeityLink** [comment]
We already had a trange() proc used to bypass the hardcoded range() limit (which basically called block() )

But since the turfs were listed from bottom to top, you'd run into this issue:

issue

The turfs south of the explosion would be properly protected from the explosion by the obstacles, while turfs north of the explosion that might seem protected at first sight are in fact not because their protection might get destroyed by the time their turf gets processed.

Now that the turfs are listed in a spiral going inwards, you're guarranted that this wall will protect you from the explosion, even if the explosion ends up destroying it.

**Comment by RemieRichards** [comment]
we have ```trange()``` too lol.

**Comment by DeityLink** [comment]
That's oldcode for you.

**Comment by Zth--** [comment]
Related to  #5746
